### ISSUE TITLE
Stackoverflow

### ISSUE THREAD

**Issue by AshotN** [lare96/luna#44]
I tried running the server and got this

`08 Feb 2017 23:58:27 io.luna.game.plugin.PluginBootstrap [LunaInitializationThread]
ERROR: Catching
java.lang.StackOverflowError: null
	at scala.tools.nsc.typechecker.Contexts$Context.bufferErrors(Contexts.scala:332) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Contexts$Context.reportErrors(Contexts.scala:333) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:672) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4524) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4580) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typedInAnyMode$1(Typers.scala:5343) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5360) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:698) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5396) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedInternal(Typers.scala:5423) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5370) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5374) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typedQualifier(Typers.scala:5472) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typedQualifier(Typers.scala:5480) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:644) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:698) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5396) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedInternal(Typers.scala:5423) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5370) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5374) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$99.apply(Typers.scala:4525) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$99.apply(Typers.scala:4525) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:680) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4524) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4580) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typedInAnyMode$1(Typers.scala:5343) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5360) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:698) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5396) ~[scala-compiler-2.11.8.jar:?]
	at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedInternal(Typers.scala:5423) ~[scala-compiler-2.11.8.jar:?]
`

**Comment by lare96** [comment]
This is happening because of this line in the [Rational](https://github.com/lare96/luna/blob/9af164ad1e866a0229afb156c36b7f432f827332/src/main/java/io/luna/util/Rational.java#L40) class for some reason. I moved [these](https://github.com/lare96/luna/blob/master/src/main/java/io/luna/game/model/Chance.java) aliases out of bootstrap.scala and instead added them into PluginBootstrap [like so](https://github.com/lare96/luna/blob/master/src/main/java/io/luna/game/plugin/PluginBootstrap.java#L202) and that seemed to fix this issue (I was able to login, all plugins were working fine, etc.)

This could mean that all I did was make the issue more volatile (sometimes it happens, sometimes it doesn't)

Does anyone have any idea how a function that retrieves the greatest common denominator of a value could have that effect? I'm going to ask around and see what I can dig up, I'm also going to do more testing tonight to see if I can reproduce it. Worst case scenario the Rational class will have to be removed (which really isn't a big deal)

OP, did you try running the server more than once?

**Comment by lare96** [comment]
Yeah OP, I can't seem to reproduce this. I've ran the server a total of 10 times and everything works as normal. Please try a few times more and see if the issue is consistent for you.

Are you using the most up-to-date master branch?

**Comment by AshotN** [comment]
From what I recall it was the latest version. I tried multiple launches. I'll check it out again when I get home

**Comment by AshotN** [comment]
It's stopped happening, then it happened once more... i restarted server and it stopped happening again -.-

I'll try to see if i can cause it to happen

**Comment by lare96** [comment]
Hmmm, so it seems this is a volatile issue. Okay, I'm going to look into it more today and I should have some sort of fix by tomorrow latest

**Comment by AshotN** [comment]
If you can't reproduce I'll let you Teamviewer onto my computer

**Comment by lare96** [comment]
Okay so I've figured out the issue, it doesn't actually seem to be the Rational class at all but maybe a compiler bug (?)

Increasing the maximum stack size via JVM arguments 'fixes' the issue but I wouldn't consider that a real fix.   Still gonna look more into it!

**Comment by lare96** [comment]
Actually, I just tried interpreting bootstrap.scala in different chunks and that seems to fix the issue as well

**Comment by AshotN** [comment]
It seems that the JVM argument does allow scala plugins to load in, but I did get some other error. I'll wait for an official patch before I worry about the new errors. BTW Is there an IRC you are on?

**Comment by lare96** [comment]
Okay an actual fix is taking awhile to figure out, so for now I'll just make it so that it prints "<file_name>.scala is too long" or something instead of just crashing with a confusing StackOverflowError

and yeah. im on the moparscape irc and i have skype as well "lare96.rsps"

**Comment by lare96** [comment]
Coming back to this with a fresh perspective, three fixes come to mind (in order of least -> most desirable and easiest -> hardest)

- Increasing the JVM stack size
- Decoupling bootstrap.scala to run in "phases" (bindings -> conversions -> monkey patching, etc.)
- Taking a look at the compiler code to see exactly why a stack overflow is occurring (not a guaranteed fix)

**Comment by ESchouten** [comment]
-Xss256m as VM argument did the trick for me.

**Comment by tlf30** [comment]
It is very, VERY unusual for an application this small to overflow the main jvm stack. Some other issue is occurring in the code causing too much data to be pushed. Possible a recursive call? Increasing the runtime stack size is just masking the problem.

**Comment by tonymorris** [comment]
It's actually very common for a small program to overflow the stack. All you need is a bit of general recursion. Scala does not to proper tail call elimination, only direct tail call elimination. The only JVM that does tail call elimination is the IBM implementation, and that is direct as well.

In this case, the `gcd` is general recursive, without a tail call. This will overflow the stack. Rewrite the function so that `gcd` is in direct tail position.

**Comment by tlf30** [comment]
@tonymorris, I would like to point out that @lare96  has already confirmed that it is not an issue with Rational, specifically the `gcd(int, int)` in `io.luna.util.Rational`? Some simple testing to confirm: commenting out that line and replacing the references to it with 1, still causes the stack overflow. 
Also I do not see how it is recursive.  Look at the javap output:
``` 
private int gcd(int, int);
    Code:
       0: iload_1
       1: i2l
       2: invokestatic  #24                 // Method java/math/BigInteger.valueOf:(J)Ljava/math/BigInteger;
       5: astore_3
       6: iload_2
       7: i2l
       8: invokestatic  #24                 // Method java/math/BigInteger.valueOf:(J)Ljava/math/BigInteger;
      11: astore        4
      13: aload_3
      14: aload         4
      16: invokevirtual #25                 // Method java/math/BigInteger.gcd:(Ljava/math/BigInteger;)Ljava/math/BigInteger;
      19: invokevirtual #26                 // Method java/math/BigInteger.intValue:()I
      22: ireturn
```

@lare96 I would also like to take note that a looped approach for this function such as (I think you had something like this before) 
``` 
private int gcd(int num0, int num1) {
        if (num0 == 0) {
            return num1;
        }

        while (num1 != 0) {
            if (num0 > num1) {
                num0 = num0 - num1;
            } else {
                num1 = num1 - num0;
            }
        }

        return num0;
    }
```
The stack overflow still occurs (as I said, I do not think the issue is here), and is also non-recursive. Also it is possible that this approach is faster as it does not have any object creation, but I have not tested that to confirm. 

@tonymorris, I think you misunderstood me. I was trying to say that one of the only times a small program should have a stack overflow on the calling stack is when a recursive call occurs. I have worked on much larger enterprise java applications where the calling stack is upwards of 4GB without recursion, simply because so many functions are on the stack at a time, on those applications it is common to have stack overflows on the calling stack if the JVM is not properly configured.

**Comment by tonymorris** [comment]
It seems I was looking at different code. I am unable to run it at the moment.

Looking at the stack trace properly, which I have seen many times before from the scala compiler, the ultimate cause is that the type-inferencer is crashing due to a program that is not type-checking. This can be resolved by starting again, and slowly introducing type-checking expressions, making sure to ascribe a type to every expression along the way (do not rely on the inferencer). You will end up finding the expression that is crashing the type-checker.

@tlf30 I too, have worked on these programs :)

**Comment by lare96** [comment]
@tonymorris Thanks for the insight, I'm going to try finding the expression(s) crashing the type-checker. Regardless though, should that not be reported as a bug to the Scala development team?

**Comment by tonymorris** [comment]
Yes. In the early Scala days, this was a common experience, as soon as you pushed the type system just a little. However, always remember, the Scala type system is turing-complete.

https://apocalisp.wordpress.com/2011/01/13/simple-ski-combinator-calculus-in-scalas-type-system/

**Comment by lare96** [comment]
Fixed as a side effect of 39a76d4e41605a665dd56a5b90f98d4b0cb1be51
